### PR TITLE
FileReader: cache files locally

### PR DESCRIPTION
**Pull request by ghost**
Cache downloaded files for faster acces next time

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -49,6 +49,9 @@ selfdrive/modeld/_modeld
 selfdrive/modeld/_dmonitoringmodeld
 /src/
 
+tools/replay/tmp/*
```
Comment (pd0wm) [inline .gitignore:52]: Use `/tmp`

**Comment by pd0wm** [comment]
As mentioned. This needs to be cleanup up, and it needs to be really solid since it's used by all offline tools. Also needs tests.

File: tools/lib/tests/test_caching.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env python3
+
+import unittest
+from tools.lib.url_file import URLFile
+
+class TestFileDownload(unittest.TestCase):
+
+  def test_downloads(self):
+    small_file_url = "https://raw.githubusercontent.com/commaai/openpilot/master/SAFETY.md"
```
Comment (adeebshihadeh) [inline tools/lib/tests/test_caching.py:9]: ideally the test shouldn't need internet. you can make a temp file, fill it with junk bytes, and refer to it like this `file:///path/to/the/tmp/file`

Comment (pd0wm) [reply tools/lib/tests/test_caching.py:9]: The tests also need to check the partial download functionality using range headers. If we want to run locally we need something like flask. However we already ran into some issues with azure implementing the range headers in a non standard way.

Comment (grekiki) [reply tools/lib/tests/test_caching.py:9]: Yeah not sure there is any benefit from not needing internet, there are some bugs in previous version that are imo not reproducible in local testing, for example getting the whole file if the file is large enough and the end of range is not specified

**Comment by grekiki** [comment]
Despite all checks working I maybe still perhaps want to run the branch on a device, just to make sure the master is not broken in some hard to find way?

**Comment by adeebshihadeh** [comment]
These tools aren't meant to run on devices. Can you revert the changes to the submodules?

**Comment by grekiki** [comment]
Yup as soon as I figure out how :)

File: tools/lib/tests/test_caching.py (hunk 2 above)
Comment (gregjhogan) [reply tools/lib/tests/test_caching.py:9]: can we change it to a qlog file to make the egress bandwidth as small as possible?
also azure storage account emulator (that runs in docker): https://github.com/Azure/Azurite

**Review by pd0wm — CHANGES_REQUESTED** [review]
General comments:
 - If we want to download more than a chunk (say the whole file), it still gets downloaded in smaller chunks. Which could mean a lot of round trips to the server. If we request a range larger than a chunk size, just round it up, download the whole thing and then write it out to multiple cache files.
  - Only use the threadlocal curl. The get_curl function was for my multithreaded implementation. That can be removed.
 - Remove commented out print statements. In general keep the code clean. This code will be read by a lot of people in the future.
 - Fix submodules.

File: tools/lib/url_file.py [hunk 3]
```diff
@@ -33,10 +39,67 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+      file_name = sha256((self._url.split("?")[0] + " " + str(increment)).encode('utf-8')).hexdigest()
+      full_path = "/tmp/" + str(file_name)
+      #If we don't have a file, download it
+      if not os.path.exists(full_path):
+        #print("Downloading")
+        data = self.read_aux(ll = CACHE_SIZE)
+        with open(full_path, "wb") as new_cached_file:
+          new_cached_file.write(data)
+
+      with open(full_path, "rb") as cached_file:
```
[the remark this answers is not in the document]
Comment (grekiki) [reply tools/lib/url_file.py:86]: I thought that would be nicer code, will change

File: tools/lib/url_file.py [hunk 4]
```diff
@@ -34,9 +47,65 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+    if self._length is not None:
+      return self._length
+    path_to_name = PATH + LENGTHS_SUBFOLDER + str(sha256((self._url.split("?")[0]).encode('utf-8')).hexdigest()) + "_length"
+    if os.path.exists(path_to_name):
+      with open(path_to_name, "r") as file_length:
+          content = file_length.read()
+          self._length = int(content)
+          return self._length
+
+    c = pycurl.Curl()
```
Comment (pd0wm) [inline tools/lib/url_file.py:60]: Reuse `self._curl`

File: tools/lib/url_file.py [hunk 5]
```diff
@@ -34,9 +47,65 @@ def __exit__(self, exc_type, exc_value, traceback):
       self._local_file = None
 
   @retry(wait=wait_random_exponential(multiplier=1, max=5), stop=stop_after_attempt(3), reraise=True)
+  def get_length(self):
+    if self._length is not None:
+      return self._length
+    path_to_name = PATH + LENGTHS_SUBFOLDER + str(sha256((self._url.split("?")[0]).encode('utf-8')).hexdigest()) + "_length"
```
Comment (pd0wm) [inline tools/lib/url_file.py:53]: - Don't put lengths in a subfolder
 - Use `os.path.join` to add subpath folders together.

File: tools/lib/url_file.py [hunk 6]
```diff
@@ -6,23 +6,36 @@
...
+K=1000
+CHUNK_SIZE=1000 * K
 
+PATH="/tmp/comma_download_cache/"
+LENGTHS_SUBFOLDER="lengths/" 
 class URLFile(object):
   _tlocal = threading.local()
 
-  def __init__(self, url, debug=False):
+  def __init__(self, url, debug=False, cache=True):
```
Comment (pd0wm) [inline tools/lib/url_file.py:22]: - Cache needs to be off by default
 - Also have env variable to turn on caching, overriding this parameter

File: tools/lib/url_file.py (hunk 5 above)
Comment (pd0wm) [reply tools/lib/url_file.py:53]: Abstract url hashing into method. It's repeated below

File: tools/lib/url_file.py [hunk 7]
```diff
@@ -34,9 +47,65 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+    while True:
+      self._pos = position
+      chunk_number = self._pos / CHUNK_SIZE
+      file_name = str(sha256((self._url.split("?")[0]).encode('utf-8')).hexdigest()) + "_" + str(chunk_number)
+      full_path = PATH + str(file_name)
+      data = None
+      #If we don't have a file, download it
+      if not os.path.exists(full_path):
+        data = self.read_aux(ll=CHUNK_SIZE)
+        with open(full_path, "wb") as new_cached_file:
```
Comment (pd0wm) [inline tools/lib/url_file.py:92]: use `atomic_write_in_dir`

File: tools/lib/url_file.py [hunk 8]
```diff
@@ -34,9 +47,65 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+    c.setopt(pycurl.NOSIGNAL, 1)
+    c.setopt(pycurl.TIMEOUT_MS, 500000)
+    c.setopt(pycurl.FOLLOWLOCATION, True)
+    c.setopt(pycurl.URL, self._url)
+    c.setopt(c.NOBODY, 1)
+    c.perform()
+
+    length = int(c.getinfo(c.CONTENT_LENGTH_DOWNLOAD))
+    self._length = length
+    with open(path_to_name, "w") as file_length:
```
Comment (pd0wm) [inline tools/lib/url_file.py:70]: use `atomic_write_in_dir`

File: tools/lib/url_file.py [hunk 9]
```diff
@@ -6,23 +6,36 @@
...
     self._local_file = None
     self._debug = debug
-
+    #We download if we cannot cache 
+    self._force_download = not cache
     try:
       self._curl = self._tlocal.curl
     except AttributeError:
       self._curl = self._tlocal.curl = pycurl.Curl()
+    if not os.path.isdir(PATH):
```
Comment (pd0wm) [inline tools/lib/url_file.py:34]: use `mkdirs_exists_ok`

**Pull request merged**